### ISSUE TITLE
Unable to install Moodle on Oracle via CLI - only via web

### ISSUE THREAD

**Issue by mudrd8mz** [moodlehq/moodle-docker#72]
With `MOODLE_DOCKER_DB=oracle` I get:

```
mudrd8mz@glux ~/git/moodle-docker $ bin/moodle-docker-compose exec webserver php admin/cli/install_database.php --agree-license --fullname="Docker moodle" --shortname="docker_moodle" --adminpass="test" --adminemail="admin@example.com"
Default exception handler: Exception - Call to a member function is_temptable() on null Debug:
Error code: generalexceptionmessage
* line 229 of /lib/ddl/sql_generator.php: Error thrown
* line 100 of /lib/ddl/database_manager.php: call to sql_generator->table_exists()
* line 466 of /question/engine/upgrade/upgradelib.php: call to database_manager->table_exists()
* line 729 of /lib/environmentlib.php: call to quiz_attempts_upgraded()
* line 475 of /lib/environmentlib.php: call to environment_custom_checks()
* line 106 of /lib/environmentlib.php: call to environment_check()
* line 161 of /admin/cli/install_database.php: call to check_moodle_environment()

!!! Exception - Call to a member function is_temptable() on null !!!
!!
Error code: generalexceptionmessage !!
!! Stack trace: * line 229 of /lib/ddl/sql_generator.php: Error thrown
* line 100 of /lib/ddl/database_manager.php: call to sql_generator-&gt;table_exists()
* line 466 of /question/engine/upgrade/upgradelib.php: call to database_manager-&gt;table_exists()
* line 729 of /lib/environmentlib.php: call to quiz_attempts_upgraded()
* line 475 of /lib/environmentlib.php: call to environment_custom_checks()
* line 106 of /lib/environmentlib.php: call to environment_check()
* line 161 of /admin/cli/install_database.php: call to check_moodle_environment()
 !!
```

However, I can still install it by going to http://localhost:8000

Not sure if this is the moodle-docker bug or the moodle core bug?

**Comment by scara** [comment]
Hi @mudrd8mz,
it's "only" a timing issue about the availability of the Oracle instance:
```
# export MOODLE_DOCKER_DB=oracle
# bin/moodle-docker-compose up -d
Creating network "moodledocker_default" with the default driver
Creating moodledocker_db_1 ...
Creating moodledocker_selenium_1 ...
Creating moodledocker_mailhog_1 ...
Creating moodledocker_db_1
Creating moodledocker_mailhog_1
Creating moodledocker_mailhog_1 ... done
Creating moodledocker_webserver_1 ...
Creating moodledocker_webserver_1 ... done
# bin/moodle-docker-compose exec webserver php admin/cli/install_database.php --agree-license --fullname="Docker moodle" --shortname="docker_moodle" --adminpass="test" --adminemail="admin@example.com"
Default exception handler: <p>Error: Database connection failed</p>
<p>It is possible that the database is overloaded or otherwise not running properly.</p>
<p>The site administrator should also check that the database details have been correctly specified in config.php</p> Debug: ORA-12537: TNS:connection closed
Error code: dbconnectionfailed
* line 189 of /lib/dml/oci_native_moodle_database.php: dml_connection_exception thrown
* line 344 of /lib/dmllib.php: call to oci_native_moodle_database->connect()
* line 615 of /lib/setup.php: call to setup_DB()
* line 84 of /config.php: call to require_once()
* line 78 of /admin/cli/install_database.php: call to require()

!!! <p>Error: Database connection failed</p>
<p>It is possible that the database is overloaded or otherwise not running properly.</p>
<p>The site administrator should also check that the database details have been correctly specified in config.php</p> !!!

Debug info: ORA-12537: TNS:connection closed
Error code: dbconnectionfailed
Stack trace: * line 189 of /lib/dml/oci_native_moodle_database.php: dml_connection_exception thrown
* line 344 of /lib/dmllib.php: call to oci_native_moodle_database->connect()
* line 615 of /lib/setup.php: call to setup_DB()
* line 84 of /config.php: call to require_once()
* line 78 of /admin/cli/install_database.php: call to require()
# bin/moodle-docker-compose exec webserver php admin/cli/install_database.php --agree-license --fullname="Docker moodle" --shortname="docker_moodle" --adminpass="test" --adminemail="admin@example.com"
Default exception handler: <p>Error: Database connection failed</p>
<p>It is possible that the database is overloaded or otherwise not running properly.</p>
<p>The site administrator should also check that the database details have been correctly specified in config.php</p> Debug: ORA-12528: TNS:listener: all appropriate instances are blocking new connections
Error code: dbconnectionfailed
* line 189 of /lib/dml/oci_native_moodle_database.php: dml_connection_exception thrown
* line 344 of /lib/dmllib.php: call to oci_native_moodle_database->connect()
* line 615 of /lib/setup.php: call to setup_DB()
* line 84 of /config.php: call to require_once()
* line 78 of /admin/cli/install_database.php: call to require()

!!! <p>Error: Database connection failed</p>
<p>It is possible that the database is overloaded or otherwise not running properly.</p>
<p>The site administrator should also check that the database details have been correctly specified in config.php</p> !!!

Debug info: ORA-12528: TNS:listener: all appropriate instances are blocking new connections
Error code: dbconnectionfailed
Stack trace: * line 189 of /lib/dml/oci_native_moodle_database.php: dml_connection_exception thrown
* line 344 of /lib/dmllib.php: call to oci_native_moodle_database->connect()
* line 615 of /lib/setup.php: call to setup_DB()
* line 84 of /config.php: call to require_once()
* line 78 of /admin/cli/install_database.php: call to require()
# bin/moodle-docker-compose exec webserver php admin/cli/install_database.php --agree-license --fullname="Docker moodle" --shortname="docker_moodle" --adminpass="test" --adminemail="admin@example.com"
Default exception handler: Exception - Call to a member function is_temptable() on null Debug:
Error code: generalexceptionmessage
* line 229 of /lib/ddl/sql_generator.php: Error thrown
* line 100 of /lib/ddl/database_manager.php: call to sql_generator->table_exists()
* line 466 of /question/engine/upgrade/upgradelib.php: call to database_manager->table_exists()
* line 729 of /lib/environmentlib.php: call to quiz_attempts_upgraded()
* line 475 of /lib/environmentlib.php: call to environment_custom_checks()
* line 106 of /lib/environmentlib.php: call to environment_check()
* line 161 of /admin/cli/install_database.php: call to check_moodle_environment()

!!! Exception - Call to a member function is_temptable() on null !!!
!!
Error code: generalexceptionmessage !!
!! Stack trace: * line 229 of /lib/ddl/sql_generator.php: Error thrown
* line 100 of /lib/ddl/database_manager.php: call to sql_generator-&gt;table_exists()
* line 466 of /question/engine/upgrade/upgradelib.php: call to database_manager-&gt;table_exists()
* line 729 of /lib/environmentlib.php: call to quiz_attempts_upgraded()
* line 475 of /lib/environmentlib.php: call to environment_custom_checks()
* line 106 of /lib/environmentlib.php: call to environment_check()
* line 161 of /admin/cli/install_database.php: call to check_moodle_environment()
 !!
# bin/moodle-docker-compose exec webserver php admin/cli/install_database.php --agree-license --fullname="Docker moodle" --shortname="docker_moodle" --adminpass="test" --adminemail="admin@example.com"
-------------------------------------------------------------------------------
== Setting up database ==
-->System
++ Success ++
...
++ Success ++
Installation completed successfully.
```
Oracle is slow when running within a container and you need to _sleep_ at least 60s before doing something (depending on your HW).

We could add some sleep time here through a loop that checks the availability of the instance each 5s and we could enforce the Moodle code to be more error proof about the availability of the database by exiting w/ a more readable error.

HTH,
Matteo

**Comment by mudrd8mz** [comment]
Interesting, thanks for the tip Matteo

**Comment by danpoltawski** [comment]
Run `bin/moodle-docker-wait-for-db` which [does this waiting ](https://github.com/moodlehq/moodle-docker/blob/master/bin/moodle-docker-wait-for-db#L14) semi intelligently.

Thats how we do the automated tests:

**Comment by danpoltawski** [comment]
(Created a pull request for docs)

**Comment by andrewnicols** [comment]
Closed via #73.

**Comment by mudrd8mz** [comment]
Awesome, thanks guys!
